lib/core/src/Cardano/Wallet/Api/Server.hs

```diff
@@ -2193,12 +2197,116 @@ constructTransaction ctx genChange (ApiT wid) body = do
     ti = timeInterpreter (ctx ^. networkLayer)
 
 balanceTransaction
-    :: forall ctx (n :: NetworkDiscriminant). ctx
+    :: forall ctx s k (n :: NetworkDiscriminant).
+        ( ctx ~ ApiLayer s k
+        , HasNetworkLayer IO ctx
+        , WalletKey k
+        , Typeable s
+        , Typeable n
+        , HardDerivation k
+        , IsOurs s Address
+        , GenChange s
+        , Bounded (Index (AddressIndexDerivationType k) 'AddressK)
+        )
+    => ctx
+    -> ArgGenChange s
     -> ApiT WalletId
     -> ApiBalanceTransactionPostData n
     -> Handler (ApiConstructTransaction n)
-balanceTransaction _ctx (ApiT _wid) _body =
-    liftHandler $ throwE ErrBalanceTxNotImplemented
+balanceTransaction ctx genChange (ApiT wid) body = do
+    pp <- liftIO $ NW.currentProtocolParameters nl
+
+    --TODO  deal with coll and validity and delegations
+    let (Tx _id _fee _coll _inps outs wdrlMap mdM _validity) = txIncoming
+
+    ((FeeEstimation feeMin _), unsignedtx) <- withWorkerCtx ctx wid liftE liftE $ \wrk -> do
+
+        (acct, _, _) <- liftHandler $ W.readRewardAccount @_ @s @k @n wrk wid
+        (wdrl, _) <-
+            if Map.member acct wdrlMap then
+                mkRewardAccountBuilder @_ @s @_ @n ctx wid (Just SelfWithdrawal)
+            else
+                mkRewardAccountBuilder @_ @s @_ @n ctx wid Nothing
+        let executionFee = calcScriptExecutionCost tl pp sealedTxIncoming
+        ttl <- liftIO $ W.getTxExpiry ti Nothing
+        let txCtx = defaultTransactionCtx
+                { txPlutusScriptExecutionCost = executionFee
+                , txMetadata = mdM
+                , txWithdrawal = wdrl
+                , txTimeToLive = ttl
+                }
+
+        (internalUtxoAvailable, wallet, pendingTxs) <-
+            liftHandler $ W.readWalletUTxOIndex @_ @s @k wrk wid
+
+        let externalSelectedUtxo =
+                UTxOIndex.fromSequence externalInputs
+
+        let utxoAvailable = UTxOSelection.fromIndexPair
+                (internalUtxoAvailable, externalSelectedUtxo)
+
+        let runSelection = W.selectAssets @_ @s @k wrk W.SelectAssetsParams
+                { outputs = outs
+                , pendingTxs
+                , txContext = txCtx
+                , utxoAvailableForInputs = utxoAvailable
+                , utxoAvailableForCollateral =
+                    UTxOIndex.toUTxO internalUtxoAvailable
+                , wallet
+                } getFee
+              where getFee = const (selectionDelta TokenBundle.getCoin)
+
+        let transform s sel' =
+                W.assignChangeAddresses genChange sel' s
+                & uncurry (W.selectionToUnsignedTx (txWithdrawal txCtx))
+        unsignedtx <- liftHandler $ W.selectAssets @_ @s @k wrk W.SelectAssetsParams
```

Comment:
Because `selectAssets` only knows about the things we tell it, and not the entire `sealedTxIncoming`, I think exotic things like submitting update proposals will result in imbalanced txs.

In a re-thought approach, related to ADP-891, and I think which you've mentioned, we could calculate `minfee sealedTxIncoming` instead and go from there (either by overestimating for the TxExtraBodyContent, or trying to find some fix point)

But I guess the similar and more pressing issue is datums, which everyone was aware of on today's call.

Reply:
agree here, it is reasonable to start with minFee, and then increment knowing the cost per inp/out/coll. But we need to remember that evaluation of min fee requires number of witnesses, so it would influence selecting inputs/collaterals in feedback loop. It is not staightforward for me how, in details, it should work